rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,243 @@
+
+![Graph of deploy lag (in days)](rfc-128/chart.svg)
+
+A backlog of deployments also makes diagnostics harder:
+
+- ⚠ We can mistakenly debug an issue against the undeployed version of the code.
+- ⚠ We may struggle to isolate a breaking change, if many were deployed together.
+
+### 2. We freely merge PRs with breaking changes
+
+We tend to merge PRs and only later worry if they break the app. This can cause the codebase to become undeployable, and interrupt other developers testing changes on Integration.
+
+Example: https://github.com/alphagov/email-alert-api/commit/8e633fc7d89d7f13bc0e5bdcab14ed3dea968289#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R20
+
+Fix: https://github.com/alphagov/email-alert-api/pull/1346
+
+I couldn't find a good example of myself doing this, but I've definitely done it too. We will never write unbreakable code, yet we have too little caution around merging our changes.
+
+### 3. We do not have a working definition of "OK"
+
+It's easy for each developer to have their own standards, and rely on their own judgement when they check that a change is safe to deploy. Theoretically, we assume a developer will:
+
+- Check the results of [the smoke tests](https://deploy.integration.publishing.service.gov.uk/job/Smokey/).
+- Look for any Icinga alerts related to the app.
+- Check the [deployment dashboard for the app](https://docs.publishing.service.gov.uk/manual/deployment-dashboards.html).
+- Generally make sure the change is tested.
+
+We often defer to automated tests to tell us when something is broken, but it's unclear whether this is really justified: there is no minimum standard for testing of GOV.UK apps.
+
+## Proposal
+
+The obvious solution to the backlog of deployments is to make them automatic. This has already begun, as part of [a trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+However, we have been unable to make further progress due to the other two issues: we are concerned that enabling automatic deployments en masse would lead to an increase in defects reaching production.
```

Comment:
One last annoying nitpick 😅 

I'm not sure what the other "two" issues are here - there's only one concern after the colon.

I suggest:

```suggestion
However, we have been unable to make further progress due to the concern that enabling automatic deployments en masse would lead to an increase in defects reaching production.
```

Reply:
OK. I'll accept one more nitpick. I've pushed [a new commit](https://github.com/alphagov/govuk-rfcs/pull/128/commits/247a09a02ac6e3cacb7a5792da0c1b2b9602c758) to clarify it - slightly different to your suggestion, as I want to try and reference the previous section.

Reply:
Ah, I see.

I think it would be easier to follow if the previous paragraph made it more obvious that it's referring back to one of the above issues. At the moment the reader needs to hold a lot of context in their head when they get to this "other two issues" statement, and it broke the flow for me.

I suggest changing this:

> The obvious solution to the backlog of deployments is to make them automatic

To this:

> The first issue (we have a constant backlog of deployments) has an obvious solution - make deployments automatic.

That provides a bit of a hint that when the next paragraph talks about "other issues" it's contrasting them with the "first issue" here. Otherwise you have to memorise the issues before you read this paragraph to be able to unpick what it means.

Sorry - this is totally a commentary on writing style, and has no bearing on the substance of the discussion 🤦.

Reply:
Sure, I did think that when I made the commit, so I've amended it as you suggested 👍. Can I resolve this now?

Reply:
Yep! Thanks!